### DISCUSSION TITLE
Custom report to show values

### DISCUSSION THREAD

**Discussion by tiagommmelo** [GothenburgBitFactory/taskwarrior#3136] [Q&A]
Hello community. I've tried Emacs and other tools, but I'm ending up coming back to taskwarrior (there's no place like home/taskwarrior). I'm just too familiar already with TW and I believe it is still the best for my workflow; it simply works and I still have my old configuration saved (which is a plus).

It turns out that I want to get my TW setup even better, so I'm changing the way I used to do stuff on the **tags**. I'll give an example:
Let's say that I have a task of someone trying to fix the computer, but that depends on a technician on site and the end user was not clear on what is wrong with the computer. I'd enter the task like this:

`task add "Fix computer problem" +@John +S_Leila +R_Bob`

In my example above:
- all tags preceding with '@' is a person that I'm waiting for something (let's say I'm waiting for John's answer on what actually is the issue with the computer);
- 'S' is a stakeholder tag; anyone involved fits here (Leila is the technician to whom I should delegate the task in this case);
- 'R' is the requestor tag; in this example, the computer belongs to Bob (a director who asked John to open a ticket with me)

So far so good

Now, an extra tool I use is TMUX so I can have command line frames around and I would have a small frame somewhere to show me all the tags with the command below:

`while true; do clear && task tags; sleep 5;done`

What this does in tmux is: clear the frame and insert the command task tags each 5 seconds.

The thing is I have a big frame with lots of tags... I can scroll, no problem, to find the stakeholders (imagine I have 50 stakeholders, 30 people waiting for and 70 requestors, this is insane)

Now, on my new setup, what I'm trying to achieve is **generating the same report, but with UDAs**. Here's a portion of my configuration file:

`
uda.req.type=string
uda.req.label=Requestor
uda.stk.type=string
uda.stk.label=Stakeholder
uda.wtg.type=string
uda.wtg.label=Waiting
`
So:
- req replaces the tags +R_SomeName
- stk replaces the tags +S_SomeName
- wtg replaces the tags +@_SomeName

With this config, I can use the tags for the contexts as it should be. For example, I have a tag +office that means that this task can only be performed while I'm working at the office, so if I'm working from home I'd filter out that task or setting it to wait until the day 'x' when I'll be at the office. I hope this makes sense.

Now, finally, here's what I'm trying to accomplish.
I know that there is a report that can show the values of UDA attributes, for example, so, if I have these custom priorities below, I can run `"task show uda.priority.values"` to get the values I've defined.

`uda.priority.values=1,2,3,,5`

Let's say that I have some tasks and some Stakeholders (stk) defined already. They are values not predefined in my configuration, but even so I tried to do this to get some results (it failed):

`task show uda.stk.values`

Before I make my question, let's say that I have 4 tasks created like this:
- task add "task 1" stk:Jonas
- task add "task 2" stk:Jonas
- task add "task 3" stk:Bob
- task add "task 4" stk:Patrick

So, my question (finally) is: **"Is there a way to run a report, list or custom report that'll show me all the values of an UDA and the count of repeated values in a way that, running the report supposing that I only have these 3 unfinished tasks above?"** The output should be something like this:

Stakeholder    Count
Bob                 1
Jonas              2
Patrick            1

I hope that question makes sense and also that this is doable somehow... Thanks in advance

**Comment by dathanb** [comment]
I don't know whether there's a way to do this with reports (I'm not a member of the taskwarrior team). But for custom things like this, I tend to lean on `task export`. For instance, to get a unique count of values for the `priority` attribute, you could pipe export to `jq`:
```
t +READY export | jq '[.[] | .priority | select(. != null)] | reduce .[] as $val ({}; .[$val] += 1)'
```
and get
```
{
  "L": 62,
  "M": 16,
  "H": 6
}
```
(for my current taskwarrior database). 

Or for more custom formatting, you can pipe your export to python. e.g., 
```
#!/usr/bin/env python3
import json
import csv
import sys

def main():
    data = sys.stdin.read().strip()

    decoder = json.JSONDecoder()

    export_data, offset = decoder.raw_decode(data)

    # count distinct priorities
    priority_counts = dict()
    for task in export_data:
        if 'priority' in task:
            priority = task['priority']
            if priority not in priority_counts:
                priority_counts[priority] = 0

            priority_counts[priority] = priority_counts[priority] + 1

    # convert to a table so we can prepend our header row
    priority_counts_table = [[priority, priority_counts[priority]] for priority in priority_counts.keys()]
    priority_counts_table.sort()
    priority_counts_table.insert(0, ["priority", "count"])

    print(table(priority_counts_table))
    
# copied from https://python-forum.io/thread-11264-post-51240.html#pid51240 for expediency 
def table(rows):
    max_widths = []
    for column in zip(*rows):
        max_widths.append(max([len(str(text)) for text in column]))
    template = '  '.join(['{{:<{}}}'.format(width) for width in max_widths])
    return '\n'.join([template.format(*row) for row in rows])


if __name__ == "__main__":
    main()
```
and `task +READY export | report.py` would generate something like
```
priority  count
H         6
L         62
M         16
```

**Comment by tiagommmelo** [comment]
Hello, thanks for your reply. The export command would help me if I could use it in a way that I can export all values that are set to an UDA. So, if my UDA is 'stk' and in various tasks I've set some people's names, then the desired output would be a list of 'stk' values with the count of how many times it repeats. So, if "Jason" is a value that was set on 3 tasks on 'stk' UDA and "Bob" was set on 5 tasks, the output would be something like this:

stk        count
Bob      5
Jason   3

So, I believe there should be a way (or it might there be a way already) to have something like this (pseudo code):

`task export /UDA:stk.value/ && /UDA:stk.value.count/`

This is only an example and of course does not work; I'm just trying to clarify what I want to achieve. 

Thanks again
